File: apps/ios/Voiid/Voiid/Main/Clips/CreatorHandleSheet.swift
//
//  CreatorHandleSheet.swift
//  Voiid
//
//  THE GATE — a creator profile is required before a first clip can be posted, and this is
//  where it gets created: on demand, at first post, not at signup (see 029's header for why
//  manufacturing a public identity for every account is both a privacy and a namespace
//  problem).
//
//  ── THIS IS PUBLIC, AND THE COPY SAYS SO ─────────────────────────────────────────
//  A creator handle is BROADCAST IDENTITY: visible to strangers, attached to every clip.
//  It is NOT the chat @username, which is half a private credential (username + PIN opens a
//  message request). They share one namespace so that a single @name can never mean two
//  different people, but they are different things and the sheet must not imply otherwise.
//

import SwiftUI

struct CreatorHandleSheet: View {
    @EnvironmentObject var creators: CreatorEngine
    @Environment(\.dismiss) private var dismiss

    /// Called with the created profile once the gate is satisfied. The caller resumes
    /// whatever it was trying to do (posting a clip).
    var onCreated: (CreatorService.Profile) -> Void

    @State private var handle = ""
    @State private var displayName = ""
    @State private var bio = ""
    @State private var state: CreatorEngine.HandleState = .idle
    @State private var submitting = false
    @State private var errorText: String?
    @FocusState private var handleFocused: Bool

    /// Submission needs only a well-formed handle. The availability check is ADVISORY —
    /// blocking on `.available` would strand the user whenever the check itself failed,
    /// and the create call re-validates under the real unique constraint regardless.
    private var canSubmit: Bool {
        !submitting && CreatorEngine.isWellFormed(normalized) && state != .taken
    }

    private var normalized: String {
        handle.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    }

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: VoiidSpacing.lg) {
                    header
                    handleField
                    displayNameField
                    bioField
                    publicNotice
                    if let errorText {
                        Text(errorText)
                            .font(VoiidFont.footnote)
                            .foregroundColor(VoiidColor.error)
                    }
                    VoiidPrimaryButton(title: submitting ? "Creating…" : "Create profile",
                                       enabled: canSubmit) {
                        Haptics.tap()
                        Task { await submit() }
                    }
                    .padding(.top, VoiidSpacing.sm)
                }
                .padding(VoiidSpacing.md)
            }
            .background(VoiidColor.background.ignoresSafeArea())
            .navigationTitle("Choose your handle")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") { dismiss() }
                        .foregroundColor(VoiidColor.textSecondary)
                }
            }
            .onAppear { handleFocused = true }
            .onDisappear { creators.cancelHandleCheck() }
        }
        .interactiveDismissDisabled(submitting)
    }

    // MARK: - Pieces

    private var header: some View {
        VStack(alignment: .leading, spacing: VoiidSpacing.xs) {
            Text("Pick a name for your clips")
                .font(VoiidFont.rounded(24, .bold))
                .foregroundColor(VoiidColor.textPrimary)
            Text("This is how people find and follow you on Clips.")
                .font(VoiidFont.subhead)
                .foregroundColor(VoiidColor.textSecondary)
        }
    }

    private var handleField: some View {
        VStack(alignment: .leading, spacing: VoiidSpacing.xs) {
            HStack(spacing: 0) {
                Text("@")
                    .font(VoiidFont.body)
                    .foregroundColor(VoiidColor.textSecondary)
                    .padding(.leading, VoiidSpacing.md)
                TextField("", text: $handle, prompt:
                    Text("handle").foregroundColor(VoiidColor.placeholder))
                    .font(VoiidFont.body)
                    .foregroundColor(VoiidColor.textPrimary)
                    .textInputAutocapitalization(.never)
                    .autocorrectionDisabled()
                    .focused($handleFocused)
                    .padding(.horizontal, VoiidSpacing.sm)
                statusIcon
                    .padding(.trailing, VoiidSpacing.md)
            }
            .frame(height: 61)
            .background(VoiidColor.fieldFill)
            .clipShape(RoundedRectangle(cornerRadius: VoiidRadius.md, style: .continuous))
            .overlay(
                RoundedRectangle(cornerRadius: VoiidRadius.md, style: .continuous)
                    .stroke(borderColor, lineWidth: 1)
            )
            // Lowercased as you type rather than at submit, so what the field shows is
            // exactly what gets reserved — handles are case-insensitive server-side.
            .onChange(of: handle) { _, new in
                let lower = new.lowercased()
                if lower != new { handle = lower; return }
                errorText = nil
                creators.checkHandle(lower) { state = $0 }
            }

            Text(hint)
                .font(VoiidFont.caption)
                .foregroundColor(hintColor)
        }
    }

    @ViewBuilder
    private var statusIcon: some View {
        switch state {
        case .checking:
            ProgressView().controlSize(.small).tint(VoiidColor.textSecondary)
        case .available:
            Image(systemName: "checkmark.circle.fill").foregroundColor(VoiidColor.success)
        case .taken, .badFormat:
            Image(systemName: "xmark.circle.fill").foregroundColor(VoiidColor.error)
        case .idle, .failed:
            EmptyView()
        }
    }

    private var borderColor: Color {
        switch state {
        case .available: return VoiidColor.success
        case .taken, .badFormat: return VoiidColor.error
        default: return handleFocused ? VoiidColor.primary : VoiidColor.fieldBorder
        }
    }

    private var hint: String {
        switch state {
        case .available: return "@\(normalized) is available."
        case .taken: return "That handle is taken."
        case .badFormat, .idle:
            return "3–20 characters. Letters, numbers and underscores, starting with a letter."
        case .checking: return "Checking…"
        case .failed(let m): return m
        }
    }

    private var hintColor: Color {
        switch state {
        case .available: return VoiidColor.success
        case .taken, .badFormat: return VoiidColor.error
        default: return VoiidColor.textSecondary
        }
    }

    private var displayNameField: some View {
        VStack(alignment: .leading, spacing: VoiidSpacing.xs) {
            Text("Display name").font(VoiidFont.footnote)
                .foregroundColor(VoiidColor.textSecondary)
            VoiidTextField(placeholder: "Optional", text: $displayName)
        }
    }

    private var bioField: some View {
        VStack(alignment: .leading, spacing: VoiidSpacing.xs) {
            Text("Bio").font(VoiidFont.footnote)
                .foregroundColor(VoiidColor.textSecondary)
            VoiidTextField(placeholder: "Optional", text: $bio)
        }
    }

    /// Clips are not end-to-end encrypted, and §6 of the rebuild doc is explicit that the UI
    /// must say so plainly rather than let someone assume Clips behaves like their chats.
    private var publicNotice: some View {
        HStack(alignment: .top, spacing: VoiidSpacing.sm) {
            Image(systemName: "globe")
                .font(.system(size: 13, weight: .semibold))
                .foregroundColor(VoiidColor.textSecondary)
            Text("Your handle, profile and clips are public and are not end-to-end encrypted. "
                 + "Your messages, calls and locations stay encrypted.")
                .font(VoiidFont.caption)
                .foregroundColor(VoiidColor.textSecondary)
                .fixedSize(horizontal: false, vertical: true)
        }
        .padding(VoiidSpacing.sm)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(VoiidColor.fieldFill.opacity(0.6))
        .clipShape(RoundedRectangle(cornerRadius: VoiidRadius.md, style: .continuous))
    }

    // MARK: - Submit

    private func submit() async {
        submitting = true
        errorText = nil
        defer { submitting = false }

        let name = displayName.trimmingCharacters(in: .whitespacesAndNewlines)
        let b = bio.trimmingCharacters(in: .whitespacesAndNewlines)
        do {
            let profile = try await creators.createProfile(
                handle: normalized,
                displayName: name.isEmpty ? nil : name,
                bio: b.isEmpty ? nil : b,
                linkURL: nil)
            Haptics.tap()
            onCreated(profile)
            dismiss()
        } catch let e as APIError {
            // 409 is the race this sheet cannot prevent: the advisory check said free, and
            // somebody took the name in between. Surfaced on the field, not as a generic
            // failure, so the fix (pick another) is obvious.
            if case .http(let status, let message, _) = e {
                switch status {
                case 409: state = .taken; errorText = "That handle was just taken. Try another."
                case 400: state = .badFormat; errorText = message
                default: errorText = message
                }
            } else {
                errorText = e.localizedDescription
            }
        } catch {
            errorText = error.localizedDescription
        }
    }
}

File: apps/android/app/src/main/java/com/voiid/app/main/clips/CreatorHandleSheet.kt
package com.voiid.app.main.clips

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Public
import androidx.compose.material.icons.filled.Cancel
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.unit.dp
import com.voiid.app.model.CreatorStore
import com.voiid.app.net.ApiError
import com.voiid.app.net.CreatorService
import com.voiid.app.ui.components.LocalVoiidHaptics
import com.voiid.app.ui.theme.VoiidColor
import com.voiid.app.ui.theme.VoiidFont
import com.voiid.app.ui.theme.VoiidRadius
import com.voiid.app.ui.theme.VoiidSpacing
import kotlinx.coroutines.launch

/**
 * THE GATE — a creator profile is required before a first clip can be posted, and this is
 * where it gets created: on demand, at first post, not at signup (see 029's header for why
 * manufacturing a public identity for every account is both a privacy and a namespace
 * problem). Mirrors iOS `CreatorHandleSheet.swift`.
 *
 * ── THIS IS PUBLIC, AND THE COPY SAYS SO ─────────────────────────────────────────
 * A creator handle is BROADCAST IDENTITY: visible to strangers, attached to every clip.
 * It is NOT the chat @username, which is half a private credential (username + PIN opens a
 * message request). They share one namespace so that a single @name can never mean two
 * different people, but they are different things and the sheet must not imply otherwise.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CreatorHandleSheet(
    creators: CreatorStore,
    onCreated: (CreatorService.Profile) -> Unit,
    onDismiss: () -> Unit,
) {
    val haptics = LocalVoiidHaptics.current
    val scope = rememberCoroutineScope()

    var handle by remember { mutableStateOf("") }
    var displayName by remember { mutableStateOf("") }
    var bio by remember { mutableStateOf("") }
    var submitting by remember { mutableStateOf(false) }
    var errorText by remember { mutableStateOf<String?>(null) }

    val state = creators.handleState
    val normalized = handle.trim().lowercase()

    // Submission needs only a well-formed handle. The availability check is ADVISORY —
    // blocking on Available would strand the user whenever the check itself failed, and the
    // create call re-validates under the real unique constraint regardless.
    val canSubmit = !submitting &&
        CreatorService.isWellFormed(normalized) &&
        state !is CreatorStore.HandleState.Taken

    LaunchedEffect(Unit) { creators.resetHandleState() }

    com.voiid.app.ui.components.VoiidSheet(
        visible = true,
        onDismiss = { if (!submitting) onDismiss() },
        detents = listOf(com.voiid.app.ui.components.VoiidDetent.Medium),
    ) {
        Column(
            Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = VoiidSpacing.md)
                .padding(bottom = VoiidSpacing.lg)
                .imePadding()
                .navigationBarsPadding(),
            verticalArrangement = Arrangement.spacedBy(VoiidSpacing.md),
        ) {
            Text(
                "Pick a name for your clips",
                style = VoiidFont.rounded(24, FontWeight.Bold),
                color = VoiidColor.textPrimary,
            )
            Text(
                "This is how people find and follow you on Clips.",
                style = VoiidFont.rounded(15),
                color = VoiidColor.textSecondary,
            )

            // ── Handle ────────────────────────────────────────────────────────────
            val borderColor = when (state) {
                is CreatorStore.HandleState.Available -> VoiidColor.success
                is CreatorStore.HandleState.Taken,
                is CreatorStore.HandleState.BadFormat -> VoiidColor.error
                else -> VoiidColor.fieldBorder
            }
            Row(
                Modifier
                    .fillMaxWidth()
                    .height(61.dp)
                    .clip(RoundedCornerShape(VoiidRadius.md))
                    .background(VoiidColor.fieldFill)
                    .border(1.dp, borderColor, RoundedCornerShape(VoiidRadius.md))
                    .padding(horizontal = VoiidSpacing.md),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text("@", style = VoiidFont.rounded(17), color = VoiidColor.textSecondary)
                Box(Modifier.weight(1f).padding(start = VoiidSpacing.xs)) {
                    BasicTextField(
                        value = handle,
                        // Lowercased as you type rather than at submit, so what the field
                        // shows is exactly what gets reserved — handles are case-insensitive
                        // server-side.
                        onValueChange = {
                            handle = it.lowercase()
                            errorText = null
                            creators.checkHandle(handle)
                        },
                        singleLine = true,
                        textStyle = VoiidFont.rounded(17).copy(color = VoiidColor.textPrimary),
                        cursorBrush = SolidColor(VoiidColor.primary),
                        keyboardOptions = KeyboardOptions(
                            capitalization = KeyboardCapitalization.None,
                            autoCorrectEnabled = false,
                        ),
                        modifier = Modifier.fillMaxWidth(),
                    )
                    if (handle.isEmpty()) {
                        Text("handle", style = VoiidFont.rounded(17), color = VoiidColor.placeholder)
                    }
                }
                when (state) {
                    is CreatorStore.HandleState.Checking ->
                        CircularProgressIndicator(
                            Modifier.size(18.dp),
                            strokeWidth = 2.dp,
                            color = VoiidColor.textSecondary,
                        )
                    is CreatorStore.HandleState.Available ->
                        Icon(Icons.Filled.CheckCircle, null,
                            tint = VoiidColor.success, modifier = Modifier.size(20.dp))
                    is CreatorStore.HandleState.Taken,
                    is CreatorStore.HandleState.BadFormat ->
                        Icon(Icons.Filled.Cancel, null,
                            tint = VoiidColor.error, modifier = Modifier.size(20.dp))
                    else -> Unit
                }
            }

            val hint = when (state) {
                is CreatorStore.HandleState.Available -> "@$normalized is available."
                is CreatorStore.HandleState.Taken -> "That handle is taken."
                is CreatorStore.HandleState.Checking -> "Checking…"
                is CreatorStore.HandleState.Failed -> state.message
                else ->
                    "3–20 characters. Letters, numbers and underscores, starting with a letter."
            }
            val hintColor = when (state) {
                is CreatorStore.HandleState.Available -> VoiidColor.success
                is CreatorStore.HandleState.Taken,
                is CreatorStore.HandleState.BadFormat -> VoiidColor.error
                else -> VoiidColor.textSecondary
            }
            Text(hint, style = VoiidFont.rounded(12), color = hintColor)

            LabelledField("Display name", displayName, "Optional") { displayName = it }
            LabelledField("Bio", bio, "Optional") { bio = it }

            // Clips are not end-to-end encrypted, and §6 of the rebuild doc is explicit that
            // the UI must say so plainly rather than let someone assume Clips behaves like
            // their chats.
            Row(
                Modifier
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(VoiidRadius.md))
                    .background(VoiidColor.fieldFill.copy(alpha = 0.6f))
                    .padding(VoiidSpacing.sm),
                horizontalArrangement = Arrangement.spacedBy(VoiidSpacing.sm),
            ) {
                Icon(Icons.Filled.Public, null,
                    tint = VoiidColor.textSecondary, modifier = Modifier.size(16.dp))
                Text(
                    "Your handle, profile and clips are public and are not end-to-end " +
                        "encrypted. Your messages, calls and locations stay encrypted.",
                    style = VoiidFont.rounded(12),
                    color = VoiidColor.textSecondary,
                )
            }

            errorText?.let {
                Text(it, style = VoiidFont.rounded(13), color = VoiidColor.error)
            }

            Box(
                Modifier
                    .fillMaxWidth()
                    .height(64.dp)
                    .clip(RoundedCornerShape(VoiidRadius.lg))
                    .background(
                        if (canSubmit) VoiidColor.primary
                        else VoiidColor.primary.copy(alpha = 0.5f)
                    )
                    .let { m ->
                        // A plain clickable, not bouncyClickable — that helper scales to 1.4x
                        // for emoji reactions and would blow up a full-width button.
                        if (!canSubmit) m else m.clickable {
                            haptics.tap()
                            submitting = true
                            errorText = null
                            scope.launch {
                                try {
                                    val profile = creators.createProfile(
                                        handle = normalized,
                                        displayName = displayName.trim().ifEmpty { null },
                                        bio = bio.trim().ifEmpty { null },
                                        linkUrl = null,
                                    )
                                    onCreated(profile)
                                    onDismiss()
                                } catch (e: ApiError.Http) {
                                    // 409 is the race this sheet cannot prevent: the advisory
                                    // check said free, and somebody took the name in between.
                                    // Surfaced on the field, not as a generic failure, so the
                                    // fix (pick another) is obvious.
                                    when (e.status) {
                                        409 -> {
                                            creators.markHandleTaken()
                                            errorText = "That handle was just taken. Try another."
                                        }
                                        else -> errorText = e.message
                                    }
                                } catch (e: Exception) {
                                    errorText = e.message ?: "Couldn't create your profile."
                                } finally {
                                    submitting = false
                                }
                            }
                        }
                    },
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    if (submitting) "Creating…" else "Create profile",
                    style = VoiidFont.rounded(17, FontWeight.SemiBold),
                    color = VoiidColor.textOnPrimary,
                )
            }
        }
    }
}

@Composable
private fun LabelledField(
    label: String,
    value: String,
    placeholder: String,
    onValueChange: (String) -> Unit,
) {
    Column(verticalArrangement = Arrangement.spacedBy(VoiidSpacing.xs)) {
        Text(label, style = VoiidFont.rounded(13), color = VoiidColor.textSecondary)
        Box(
            Modifier
                .fillMaxWidth()
                .height(61.dp)
                .clip(RoundedCornerShape(VoiidRadius.md))
                .background(VoiidColor.fieldFill)
                .border(1.dp, VoiidColor.fieldBorder, RoundedCornerShape(VoiidRadius.md))
                .padding(horizontal = VoiidSpacing.md),
            contentAlignment = Alignment.CenterStart,
        ) {
            BasicTextField(
                value = value,
                onValueChange = onValueChange,
                singleLine = true,
                textStyle = VoiidFont.rounded(17).copy(color = VoiidColor.textPrimary),
                cursorBrush = SolidColor(VoiidColor.primary),
                modifier = Modifier.fillMaxWidth(),
            )
            if (value.isEmpty()) {
                Text(placeholder, style = VoiidFont.rounded(17), color = VoiidColor.placeholder)
            }
        }
    }
}
